graylog2-server/src/test/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingRotationAndRetentionTest.java

```diff
@@ -0,0 +1,333 @@
+        timeBasedSizeOptimizingStrategy.rotate(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_0", "test_1", "test_2"));
+
+        // If an index reaches TimeBasedSizeOptimizingStrategy.MAX_INDEX_SIZE
+        // it can be rotated, even if the rotation period has been reached.
+        indexSet.getNewest().setSize(TimeBasedSizeOptimizingStrategy.MAX_INDEX_SIZE.toBytes());
+        clock.plus(12, TimeUnit.HOURS);
+        timeBasedSizeOptimizingStrategy.rotate(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_0", "test_1", "test_2", "test_3"));
+
+        // If an index doesn't reach TimeBasedSizeOptimizingStrategy.MAX_INDEX_SIZE
+        // But exceeds the optimization "leeway" ( indexLifetimeHard - indexLifetimeSoft) (6 - 4) = 2 days
+        // it will also be rotated.
+        indexSet.getNewest().setSize(100);
+        clock.plus(3, TimeUnit.DAYS);
+        timeBasedSizeOptimizingStrategy.rotate(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_0", "test_1", "test_2", "test_3", "test_4"));
+    }
+
+    void testRetention() {
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_0", "test_1", "test_2", "test_3", "test_4"));
+
+        LOG.info("Existing indices from rotation:\n<{}>",
+                indexSet.getIndices().stream()
+                        .map(i -> f("%s: created: %s closed: %s/(%s ago)\n", i.name, i.creationDate, i.closingDate, getBetween(i))).toList());
+
+        // Retention should happen after indexLifetimeSoft (4 days)
+        deletionRetentionStrategy.retain(indexSet);
+
+        // Only test_0 will be retained
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_1", "test_2", "test_3", "test_4"));
+
+        // Moving forward in time will retain more indices
+        clock.plus(12, TimeUnit.HOURS);
+        deletionRetentionStrategy.retain(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_2", "test_3", "test_4"));
+
+        clock.plus(1, TimeUnit.DAYS);
+        deletionRetentionStrategy.retain(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_3", "test_4"));
+
+        clock.plus(3, TimeUnit.DAYS);
+        deletionRetentionStrategy.retain(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_4"));
+
+        // Active index will never be retained
+        clock.plus(10, TimeUnit.DAYS);
+        deletionRetentionStrategy.retain(indexSet);
+        assertThat(indexSet.getIndicesNames()).isEqualTo(List.of("test_4"));
+    }
+
+    private String getBetween(TestIndex i) {
+        if (i.getClosingDate() == null) {
+            return "null";
+        }
+        return org.joda.time.Period.fieldDifference(clock.nowUTC().toLocalDateTime(), i.getClosingDate().toLocalDateTime()).toString();
+    }
+
+    class TestIndexSet extends org.graylog2.indexer.TestIndexSet {
```

Comment:
TestIndexSet and TestIndex could be useful for other tests. Maybe break out those into `indexer.indices.util`?

Reply:
hmm, not sure about that. 
it all relies on a lot of mocks (line 107)